modules/apps/headless/headless-admin-user/headless-admin-user-impl/src/main/java/com/liferay/headless/admin/user/internal/resource/v1_0/RoleResourceImpl.java

```diff
@@ -405,6 +424,22 @@ true, _getActions(serviceBuilderRole.getRoleId()),
 			serviceBuilderRole);
 	}
 
+	@Override
+	public Role putRoleByExternalReferenceCode(
+			String externalReferenceCode, Role role)
+		throws Exception {
+
+		com.liferay.portal.kernel.model.Role serviceBuilderRole =
+			_roleService.fetchRoleByExternalReferenceCode(
+				externalReferenceCode, contextCompany.getCompanyId());
+
+		if (serviceBuilderRole == null) {
+			return putRole(0L, role);
+		}
+
+		return putRole(serviceBuilderRole.getRoleId(), role);
+	}
```

Comment:
for the put I don't know because
1) if we do the put by ERC ---->
    if it exists ---> we update it
    if doesn't exist ---> we create the role but with the ERC used in the request (or better, we are creating the role with the ERC used in the body --- we check using ERC in the path but we create using the ERC in the body)
 2) if we do the put by ID (or with the 0 like in this case) ---->
    if it exists ---> we update it
    if doesn't exist ---> we create the role but without the ID used in the request

I have to think but give me your ideas about this

Reply:
@smottal I dont think it is possible to update the ID using the ID from the request body. From what I saw, this is generally not possible. ID is always generated (or incremented in the code, I saw something like this somewhere).
So what happens here, I think is that simply the following branch of the if executes (which would execute anyway):
if (serviceBuilderRole == null) {
			serviceBuilderRole = _roleService.addRole(
				role.getExternalReferenceCode(), className, 0, role.getName(),
				_getTitleMap(role), _getDescriptionMap(role), type, null,
				serviceContext);
		}
		
The role's ID is not passed to this addRole method.

Reply:
I also checked that addRole() does not have a version that has an ID param.